css/ampshared.css

```diff
@@ -237,7 +239,8 @@ i-amphtml-sizer {
 
 /* Hide all children of non-container elements. */
 .i-amphtml-notbuilt:not(.i-amphtml-layout-container) > * ,
-[layout]:not([layout=container]):not(.i-amphtml-element) > *
+[layout]:not([layout=container]):not(.i-amphtml-element) > *,
+[width][height][sizes]:not([layout]):not(.i-amphtml-element) > *
```

Comment:
I know we wanted to preserve CSS specificity to be safe, but I'm really curious whether we can simplify these by removing post-build selectors.

Reply:
/cc @dvoytenko

Reply:
Not sure I understand the question completely. But this is not so much about specificity as the lifecycle. There are three levels of selectors:

1. Pre-resolve. E.g. what would be otherwise written as `:not(:defined)`. Not doable w/o native CE, so we approximate it with `:not(.i-amphtml-element)`.
2. Resolved, but not built. This is `.i-amphtml-notbuilt`.
3. Resolved and built.

We definitely want to lose 1 and 2, once 3 happens. And the best way to do it is via `:not()` expressions. This doesn't affect specificity in this case - they simply no longer apply.

Reply:
Yea, I get that but it's the same CSS rule so why do we need (2) and (3) at all? Can we simplify:
```css
.i-amphtml-notbuilt:not(.i-amphtml-layout-container) > *,
[layout]:not([layout=container]):not(.i-amphtml-element) > *
```
into:
```css
[layout]:not([layout=container]) > *
```
Basically raising this question again (3 years later 😄): https://github.com/ampproject/amphtml/pull/12077#discussion_r151261804

Reply:
Will, maybe this is worth offline discussion, but I don't see how yet. All elements right now expect that the children will be automatically shown as soon as the element is built. If we remove `.i-amphtml-notbuilt`, we'd loose that automatic behavior, no?

Reply:
Oh sorry, I thought we had a CSS rule for post-built elements like:
```css
i-amphtml-built > * {
  display: initial;
}
```
Then perhaps we could combine (1) and (2) as I suggested.

Reply:
Yeah. With these we do run into specificity issues. E.g. someone might have:

```
<style amp-custom>
  .myclass {
     display: flex;
  }
</style>
<amp-component>
  <div class="myclass"></div>
</amp-component>
```

And we very easily fall into a specificity trap this way. Basically, the easiest way for us to avoid specificity problems (and maybe even the only way in some cases) is to reverse styling: we try to never set any particular `display` style - only unset it in some cases. Hence lot's of `:not(...) {display:none}` expressions. E.g. "normally don't touch, but in this specific hide do not display".

Reply:
I see, thanks for clarifying.